modules/cats/src/smithy4s/catz/SchemaVisitorHash.scala

```diff
@@ -0,0 +1,150 @@
+      tag: Primitive[P]
+  ): Hash[P] = primHashPf(tag)
+
+  override def collection[C[_], A](
+      shapeId: ShapeId,
+      hints: Hints,
+      tag: CollectionTag[C],
+      member: Schema[A]
+  ): Hash[C[A]] = {
+    implicit val memberHash: Hash[A] = self(member)
+    tag match {
+      case CollectionTag.ListTag       => Hash[List[A]]
+      case CollectionTag.SetTag        => Hash[Set[A]]
+      case CollectionTag.VectorTag     => Hash[Vector[A]]
+      case CollectionTag.IndexedSeqTag => Hash[Seq[A]].contramap(_.toIndexedSeq)
+    }
+  }
+
+  override def map[K, V](
+      shapeId: ShapeId,
+      hints: Hints,
+      key: Schema[K],
+      value: Schema[V]
+  ): Hash[Map[K, V]] = {
+    implicit val keyHash: Hash[K] = self(key)
+    implicit val valueHash: Hash[V] = self(value)
+    Hash[Map[K, V]]
+  }
+
+  override def enumeration[E](
+      shapeId: ShapeId,
+      hints: Hints,
+      values: List[EnumValue[E]],
+      total: E => EnumValue[E]
+  ): Hash[E] = {
+    implicit val enumValueHash: Hash[EnumValue[E]] =
+      Hash[String].contramap(_.stringValue)
+    Hash[EnumValue[E]].contramap(total)
+  }
+
+  override def struct[S](
+      shapeId: ShapeId,
+      hints: Hints,
+      fields: Vector[SchemaField[S, _]],
+      make: IndexedSeq[Any] => S
+  ): Hash[S] = {
+    new Hash[S] {
+      override def hash(x: S): Int = {
+        def forField[A2](field: Field[Schema, S, A2]): Hash[S] = {
+          val hashField: Hash[A2] =
+            field.foldK(new Field.FolderK[Schema, S, Hash]() {
+              override def onRequired[A](
+                  label: String,
+                  instance: Schema[A],
+                  get: S => A
+              ): Hash[A] = self(instance)
+
+              override def onOptional[A](
+                  label: String,
+                  instance: Schema[A],
+                  get: S => Option[A]
+              ): Hash[Option[A]] = {
+                implicit val hashA: Hash[A] = self(instance)
+                Hash[Option[A]]
+              }
+            })
+          hashField.contramap(field.get)
+        }
+        fields.map(field => forField(field)).map(hash => hash.hash(x)).sum
+      }
+      override def eqv(x: S, y: S): Boolean =
+        self.struct(shapeId, hints, fields, make).eqv(x, y)
+
+    }
+  }
+
+  override def union[U](
+      shapeId: ShapeId,
+      hints: Hints,
+      alternatives: Vector[SchemaAlt[U, _]],
+      dispatch: Alt.Dispatcher[Schema, U]
+  ): Hash[U] = {
+    val precomputed: Precompiler[Schema, Hash] = new Precompiler[Schema, Hash] {
+      override def apply[A](label: String, instance: Schema[A]): Hash[A] =
+        self(instance)
+    }
+    implicit val encoderKHash: EncoderK[Hash, Int] = new EncoderK[Hash, Int] {
+      override def apply[A](fa: Hash[A], a: A): Int = fa.hash(a)
+
+      override def absorb[A](f: A => Int): Hash[A] = new Hash[A] {
+        override def hash(x: A): Int = f(x)
+        override def eqv(x: A, y: A): Boolean = f(x) == f(y)
```

Comment:
That is very incorrect : you cannot decide on the equivalence of two values by comparing their hashcode. The implementation of `union` should look like the one of your current `Eq` related Schema Visitor (which will disappear in favour of this one).

Reply:
Thanks @Baccata 
To clarify , A Union of an Int or String should have the hashcode of the underlying value 
i.e 
```
sealed trait IntOrString
case class IntValue(i:Int) extends IntOrString
case class StringValue(str:String) extends IntOrString
```
the hash method derived should produce the hashCode of the internal data type

Reply:
Good question. I think the label should be added to the hash computation. If you have : 

``` 
union SomeInt {
  a: Int 
  b: Int
} 
```

then the hashcode should differen depending on whether you're in the `a` case or the `b` case, which implies that you may want to use the hashcode of the label in addition to the hashcode of the underlying value (by following best practice for computing hashcodes for composite values)